spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/ProposalsCollectorSupportUtils.java

```diff
@@ -72,6 +74,65 @@ void doAddProposals(String text, String startsWith, AppRegistration appRegistrat
 		}
 	}
 
+	void addValueHintsProposals(final String dsl, AppRegistration appRegistration, final List<CompletionProposal> collector, final String propertyName, final ValueHintProvider[] valueHintProviders){
+		final Resource metadataResource = appRegistration.getMetadataResource();
+
+		final URLClassLoader classLoader = metadataResolver.createAppClassLoader(metadataResource);
+		this.doWithClassLoader(classLoader, new Callback<Void>() {
+			@Override
+			public Void invoke() {
+				CompletionProposal.Factory proposals = expanding(dsl);
+				List<ConfigurationMetadataProperty> whiteList = metadataResolver.listProperties(metadataResource);
+				for (ConfigurationMetadataProperty property : metadataResolver.listProperties(metadataResource, true)) {
+					if (CompletionUtils.isMatchingProperty(propertyName, property, whiteList)) {
+						for (ValueHintProvider valueHintProvider : valueHintProviders) {
+							for (ValueHint valueHint : valueHintProvider.generateValueHints(property, classLoader)) {
+								collector.add(proposals.withSuffix(String.valueOf(valueHint.getValue()),
+										valueHint.getShortDescription()));
+							}
+						}
+					}
+				}
+				return null;
+			}
+		});
+	}
+
+	boolean addAlreadyTypedValueHintsProposals(final String text, AppRegistration appRegistration, final List<CompletionProposal> collector, final String propertyName, final ValueHintProvider[] valueHintProviders, final String alreadyTyped){
+		final Resource metadataResource = appRegistration.getMetadataResource();
+
+		final URLClassLoader classLoader = metadataResolver.createAppClassLoader(metadataResource);
+		return this.doWithClassLoader(classLoader, new Callback<Boolean>() {
```

Comment:
Also, it would be helpful if you were to review the names for all those new `add*` operations in `ProposalsCollectorSupportUtils`. I did my best to extract the _intention_ from the code, but could be wrong.